ssl/t1_lib.c

```diff
@@ -417,6 +417,315 @@ int ssl_load_groups(SSL_CTX *ctx)
+        if (sinf->sigalg_oid == NULL)
+            goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_SIG_NAME);
+    if (p == NULL)
+        sinf->sig_name = NULL;
+    else if (p->data_type != OSSL_PARAM_UTF8_STRING)
+        goto err;
+    else {
+        sinf->sig_name = OPENSSL_strdup(p->data);
+        if (sinf->sig_name == NULL)
+            goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_SIG_OID);
+    if (p == NULL)
+        sinf->sig_oid = NULL;
+    else if (p->data_type != OSSL_PARAM_UTF8_STRING)
+        goto err;
+    else {
+        sinf->sig_oid = OPENSSL_strdup(p->data);
+        if (sinf->sig_oid == NULL)
+            goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_HASH_NAME);
+    if (p == NULL)
+        sinf->hash_name = NULL;
+    else if (p->data_type != OSSL_PARAM_UTF8_STRING)
+        goto err;
+    else {
+        sinf->hash_name = OPENSSL_strdup(p->data);
+        if (sinf->hash_name == NULL)
+            goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_HASH_OID);
+    if (p == NULL)
+        sinf->hash_oid = NULL;
+    else if (p->data_type != OSSL_PARAM_UTF8_STRING)
+        goto err;
+    else {
+        sinf->hash_oid = OPENSSL_strdup(p->data);
+        if (sinf->hash_oid == NULL)
+            goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_KEYTYPE);
+    if (p == NULL)
+        sinf->keytype = NULL;
+    else if (p->data_type != OSSL_PARAM_UTF8_STRING)
+        goto err;
+    else {
+        sinf->keytype = OPENSSL_strdup(p->data);
+        if (sinf->keytype == NULL)
+            goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_KEYTYPE_OID);
+    if (p == NULL)
+        sinf->keytype_oid = NULL;
+    else if (p->data_type != OSSL_PARAM_UTF8_STRING)
+        goto err;
+    else {
+        sinf->keytype_oid = OPENSSL_strdup(p->data);
+        if (sinf->keytype_oid == NULL)
+            goto err;
+    }
+
+    /* The remaining parameters below are mandatory again */
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_MIN_TLS);
+    if (p == NULL || !OSSL_PARAM_get_int(p, &sinf->mintls)) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto err;
+    }
+    if ((sinf->mintls != 0) && (sinf->mintls != -1) &&
+        ((sinf->mintls < TLS1_VERSION))) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto err;
+    }
+
+    p = OSSL_PARAM_locate_const(params, OSSL_CAPABILITY_TLS_SIGALG_MAX_TLS);
+    if (p == NULL || !OSSL_PARAM_get_int(p, &sinf->maxtls)) {
+        ERR_raise(ERR_LIB_SSL, ERR_R_PASSED_INVALID_ARGUMENT);
+        goto err;
+    }
+    if ((sinf->maxtls != 0) && (sinf->maxtls != -1) &&
+        ((sinf->maxtls < sinf->mintls)
+         || (sinf->maxtls > TLS_MAX_VERSION_INTERNAL))) {
```

Comment:
I don't think claiming supportin maxtls for TLSv1.4 or above is an invalid argument. If this provider has been written for some future version of OpenSSL that has support for TLSv1.4 - but the provider is otherwise compatible with OpenSSL 3.2 then that should be fine. We obviously won't do anything with TLSv1.4 but we can still use the sigalg in TLSv1.3.

Reply:
Strict check removed in https://github.com/openssl/openssl/pull/19312/commits/f9b5466a5029bc2f26e97b6662857f33a89cd0e5